### PR TITLE
Fix to #1709 - Simplify pattern for Include around collections/references

### PR DESCRIPTION
**Pull request by maumar**
Added ThenInclude overload for a reference. Also added tests that were removed when the overload was removed previously.

### CODE REVIEW

**Comment by maumar** [comment]
@anpete @divega

File: src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs [hunk 1]
```diff
@@ -2185,22 +2185,41 @@ public static class EntityFrameworkQueryableExtensions
...
+            = typeof(EntityFrameworkQueryableExtensions)
+                .GetTypeInfo().GetDeclaredMethods("ThenInclude")
+                .Single(mi => mi.GetParameters().Any(pi => pi.Name == "navigationPropertyPath"));
+
 
         // TODO API docs once we resolve #1709
         public static IIncludableQueryable<TEntity, TProperty> ThenInclude<TEntity, TPreviousProperty, TProperty>(
             [NotNull] this IIncludableQueryable<TEntity, ICollection<TPreviousProperty>> source,
-            [NotNull] Expression<Func<TPreviousProperty, TProperty>> navigationPropertyPath)
+            [NotNull] Expression<Func<TPreviousProperty, TProperty>> collectionNavigationPropertyPath)
```
Comment (divega) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: The difference between the two overloads is in the source query argument, not the navprop path argument. Changing the name us very misleading.

File: src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs [hunk 2]
```diff
@@ -2185,22 +2185,41 @@ public static class EntityFrameworkQueryableExtensions
                         new[] { source.Expression, Expression.Quote(navigationPropertyPath) })));
         }
 
-        internal static readonly MethodInfo ThenIncludeMethodInfo
+        internal static readonly MethodInfo ThenIncludeCollectionMethodInfo
```
Comment (divega) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: Something like `ThenIncludeAfterCollectionMethodInfo` would probably make what this is clearer.

File: src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs [hunk 3]
```diff
@@ -2185,22 +2185,41 @@ public static class EntityFrameworkQueryableExtensions
...
             = typeof(EntityFrameworkQueryableExtensions)
                 .GetTypeInfo().GetDeclaredMethods("ThenInclude")
```
Comment (divega) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: OT: Can we start using nameof() for these?

**Comment by maumar** [comment]
@divega - updated the PR with your comments

File: src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs [hunk 4]
```diff
@@ -2185,23 +2184,41 @@ public static class EntityFrameworkQueryableExtensions
...
+
+        internal static readonly MethodInfo ThenIncludeAfterReferenceMethodInfo
+            = typeof(EntityFrameworkQueryableExtensions)
+                .GetTypeInfo().GetDeclaredMethods(nameof(EntityFrameworkQueryableExtensions.ThenInclude))
+                .Single(mi => mi.GetParameters().Any(pi => pi.Name == "sourceReference"));
 
         // TODO API docs once we resolve #1709
         public static IIncludableQueryable<TEntity, TProperty> ThenInclude<TEntity, TPreviousProperty, TProperty>(
-            [NotNull] this IIncludableQueryable<TEntity, ICollection<TPreviousProperty>> source,
+            [NotNull] this IIncludableQueryable<TEntity, ICollection<TPreviousProperty>> sourceCollection,
```
Comment (divega) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: Renaming the source parameter should be fine, but it is hard to come up with a good name. E.g. `sourceQueryIncludingCollectionNavigation` would be more accurate :stuck_out_tongue_winking_eye:. Another option would be to leave the parameter names the same and select the MethodInfos using type information, e.g. something like this: 

``` C#
internal static readonly MethodInfo ThenIncludeAfterCollectionMethodInfo =
    typeof(EntityFrameworkQueryableExtensions)
        .GetTypeInfo().GetDeclaredMethods(nameof(EntityFrameworkQueryableExtensions.ThenInclude))
        .Single(mi => mi.GetParameters()[0].ParameterType.GenericTypeArguments[1].IsGenericType);

internal static readonly MethodInfo ThenIncludeAfterReferenceMethodInfo = 
    typeof(EntityFrameworkQueryableExtensions)
        .GetTypeInfo().GetDeclaredMethods(nameof(EntityFrameworkQueryableExtensions.ThenInclude))
        .Single(mi => mi.GetParameters()[0].ParameterType.GenericTypeArguments[1].IsGenericParameter);
```

File: src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs [hunk 5]
```diff
@@ -21,7 +20,7 @@ public static class EntityFrameworkQueryableExtensions
     {
         #region Any/All
 
-        private static readonly MethodInfo _any = GetMethod("Any");
+        private static readonly MethodInfo _any = GetMethod(nameof(Queryable.Any));
```
Comment (divega) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: Thanks for trying this. It ends up looking a bit more verbose than the string literal. Let's find out what @anpete prefers.

Comment (divega) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: BTW, I like very much the way they in [Queryable.cs](https://github.com/dotnet/corefx/blob/master/src/System.Linq.Queryable/src/System/Linq/Queryable.cs) they take advantage of the compiler on every call, even if it potentially is less performant. Let's leave that to chat later.

File: src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs (hunk 4 above)
Comment (maumar) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: I will try to differentiate based on type. Proposed name seems too long to me.

Comment (divega) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: Agreed :smile: Thanks!

**Comment by divega** [comment]
Other than the feedback provided so far, looks :shipit: to me.

File: src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs (hunk 5 above)
Comment (anpete) [inline src/EntityFramework.Core/Extensions/EntityFrameworkQueryableExtensions.cs]: Like nameof. Find doing the Queryable thing too - anything that reduces boilerplate/complexity here is great.

**Comment by anpete** [comment]
:shipit:

**Pull request merged**